File: Topics/Design/OrderBook.java
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.TreeMap;

public class OrderBook {

    enum Side {
        BUY,
        SELL
    }

    static class Order {
        final long id;
        final Side side;
        final int price;
        int quantity;

        Order(long id, Side side, int price, int quantity) {
            this.id = id;
            this.side = side;
            this.price = price;
            this.quantity = quantity;
        }

        @Override
        public String toString() {
            return "Order{" +
                    "id=" + id +
                    ", side=" + side +
                    ", price=" + price +
                    ", quantity=" + quantity +
                    '}';
        }
    }

    static class OrderLocation {
        final Side side;
        final int price;

        OrderLocation(Side side, int price) {
            this.side = side;
            this.price = price;
        }
    }

    /*
     * Buy prices: highest first.
     *
     * At each price level, LinkedHashMap preserves insertion order,
     * so the first entry is the oldest order.
     */
    private final TreeMap<Integer, LinkedHashMap<Long, Order>> bids =
            new TreeMap<>((a, b) -> Integer.compare(b, a));

    /*
     * Sell prices: lowest first.
     */
    private final TreeMap<Integer, LinkedHashMap<Long, Order>> asks =
            new TreeMap<>();

    /*
     * orderId -> side and price
     *
     * Once we know the side and price, LinkedHashMap can remove the
     * order directly by ID in average O(1).
     */
    private final Map<Long, OrderLocation> orderIndex =
            new java.util.HashMap<>();

    public boolean addOrder(Order order) {
        if (order == null
                || order.id <= 0
                || order.price <= 0
                || order.quantity <= 0) {
            return false;
        }

        if (orderIndex.containsKey(order.id)) {
            return false;
        }

        if (order.side == Side.BUY) {
            matchBuyOrder(order);
        } else {
            matchSellOrder(order);
        }

        if (order.quantity > 0) {
            addToBook(order);
        }

        return true;
    }

    private void addToBook(Order order) {
        TreeMap<Integer, LinkedHashMap<Long, Order>> book =
                order.side == Side.BUY ? bids : asks;

        LinkedHashMap<Long, Order> priceLevel =
                book.computeIfAbsent(
                        order.price,
                        ignored -> new LinkedHashMap<>()
                );

        priceLevel.put(order.id, order);

        orderIndex.put(
                order.id,
                new OrderLocation(order.side, order.price)
        );
    }

    private void matchBuyOrder(Order incoming) {
        while (incoming.quantity > 0
                && !asks.isEmpty()
                && incoming.price >= asks.firstKey()) {

            Map.Entry<Integer, LinkedHashMap<Long, Order>> bestAsk =
                    asks.firstEntry();

            LinkedHashMap<Long, Order> priceLevel =
                    bestAsk.getValue();

            Map.Entry<Long, Order> restingEntry =
                    priceLevel.entrySet().iterator().next();

            Order restingOrder = restingEntry.getValue();

            int tradedQuantity =
                    Math.min(incoming.quantity, restingOrder.quantity);

            incoming.quantity -= tradedQuantity;
            restingOrder.quantity -= tradedQuantity;

            if (restingOrder.quantity == 0) {
                priceLevel.remove(restingOrder.id);
                orderIndex.remove(restingOrder.id);

                if (priceLevel.isEmpty()) {
                    asks.remove(bestAsk.getKey());
                }
            }
        }
    }

    private void matchSellOrder(Order incoming) {
        while (incoming.quantity > 0
                && !bids.isEmpty()
                && incoming.price <= bids.firstKey()) {

            Map.Entry<Integer, LinkedHashMap<Long, Order>> bestBid =
                    bids.firstEntry();

            LinkedHashMap<Long, Order> priceLevel =
                    bestBid.getValue();

            Map.Entry<Long, Order> restingEntry =
                    priceLevel.entrySet().iterator().next();

            Order restingOrder = restingEntry.getValue();

            int tradedQuantity =
                    Math.min(incoming.quantity, restingOrder.quantity);

            incoming.quantity -= tradedQuantity;
            restingOrder.quantity -= tradedQuantity;

            if (restingOrder.quantity == 0) {
                priceLevel.remove(restingOrder.id);
                orderIndex.remove(restingOrder.id);

                if (priceLevel.isEmpty()) {
                    bids.remove(bestBid.getKey());
                }
            }
        }
    }

    public boolean cancelOrder(long orderId) {
        OrderLocation location = orderIndex.get(orderId);

        if (location == null) {
            return false;
        }

        TreeMap<Integer, LinkedHashMap<Long, Order>> book =
                location.side == Side.BUY ? bids : asks;

        LinkedHashMap<Long, Order> priceLevel =
                book.get(location.price);

        if (priceLevel == null) {
            return false;
        }

        Order removedOrder = priceLevel.remove(orderId);

        if (removedOrder == null) {
            return false;
        }

        if (priceLevel.isEmpty()) {
            book.remove(location.price);
        }

        orderIndex.remove(orderId);

        return true;
    }

    /*
     * Simple cancel-and-replace implementation.
     * The modified order loses its original time priority.
     */
    public boolean modifyOrder(
            long orderId,
            int newPrice,
            int newQuantity
    ) {
        if (newPrice <= 0 || newQuantity <= 0) {
            return false;
        }

        OrderLocation location = orderIndex.get(orderId);

        if (location == null) {
            return false;
        }

        Side side = location.side;

        cancelOrder(orderId);

        return addOrder(
                new Order(
                        orderId,
                        side,
                        newPrice,
                        newQuantity
                )
        );
    }

    public Integer getBestBid() {
        return bids.isEmpty() ? null : bids.firstKey();
    }

    public Integer getBestAsk() {
        return asks.isEmpty() ? null : asks.firstKey();
    }

    public static void main(String[] args) {
        OrderBook book = new OrderBook();

        book.addOrder(new Order(1, Side.BUY, 100, 50));
        book.addOrder(new Order(2, Side.BUY, 100, 30));
        book.addOrder(new Order(3, Side.SELL, 102, 40));

        System.out.println("Best bid: " + book.getBestBid());
        System.out.println("Best ask: " + book.getBestAsk());

        /*
         * Incoming BUY 103 can match the resting SELL 102.
         */
        book.addOrder(new Order(4, Side.BUY, 103, 20));

        book.cancelOrder(2);

        /*
         * Cancel order 1 and re-add it at price 103.
         */
        book.modifyOrder(1, 103, 50);

        System.out.println("Best bid: " + book.getBestBid());
        System.out.println("Best ask: " + book.getBestAsk());
    }
}

File: Topics/Design/OrderBook.cpp
enum class Side {Buy, Sell}; // Strongly Typed Enum

    struct Order {
        std::int64_t id;
        Side side;
        int price;
        int quantity;
    }

    class OrderBook {

        private:
            std::mutex mutex_;

            using OrderQueue = std::list<Order>; // std::vector if reallocate fail

            std::map<int, OrderQueue, std::greater<int>> bids_; // buy stocks
            std::map<int, OrderQueue> asks_; // sell stocks

            struct OrderLocation {
                Side side;
                int price;
                OrderQueue::iterator iterator;
            };

            std::unordered_map<std::int64_t, OrderLocation> orderIndex_;

            void addToBook(const Order& order) {
                if (order.side == Side::Buy) {
                    auto& queue = bids_[order.price];
                    queue.push_back(order);
                    auto orderIt = std::prev(queue.end());
                    orderIndex_[order.id] = {
                        Side::Buy,
                        order.price,
                        orderIt
                    };
                } else {
                    auto& queue = asks_[order.price];
                    queue.push_back(order);
                    auto orderIt = std::prev(queue.end());
                    orderIndex_[order.id] = {
                        Side::Sell,
                        order.price,
                        orderIt
                    };
                }
            }

            void matchBuyOrder(Order& incoming) {
                while (
                    incoming.quantity > 0 &&
                    !asks_.empty() &&
                    incoming.price >= asks_.begin()->first
                ) {
                    auto bestAskIt = asks_.begin();
                    Order& restingOrder = bestAskIt->second.front();

                    int tradedQuantity =
                        std::min(incoming.quantity, restingOrder.quantity);

                    incoming.quantity -= tradedQuantity;
                    restingOrder.quantity -= tradedQuantity;

                    if (restingOrder.quantity == 0) {
                        std::int64_t filledOrderId = restingOrder.id;

                        bestAskIt->second.pop_front();
                        orderIndex_.erase(filledOrderId);

                        if (bestAskIt->second.empty()) {
                            asks_.erase(bestAskIt);
                        }
                    }
                }
            }

            void matchSellOrder(Order& incoming) {
                while (
                    incoming.quantity > 0 &&
                    !bids_.empty() &&
                    incoming.price <= bids_.begin()->first
                ) {
                    auto bestBidIt = bids_.begin();
                    Order& restingOrder = bestBidIt->second.front();

                    int tradedQuantity =
                        std::min(incoming.quantity, restingOrder.quantity);

                    incoming.quantity -= tradedQuantity;
                    restingOrder.quantity -= tradedQuantity;

                    if (restingOrder.quantity == 0) {
                        std::int64_t filledOrderId = restingOrder.id;

                        bestBidIt->second.pop_front();
                        orderIndex_.erase(filledOrderId);

                        if (bestBidIt->second.empty()) {
                            bids_.erase(bestBidIt);
                        }
                    }
                }
            }

            void matchOrder(Order& incoming) {
                if (incoming.side == Side::Buy) {
                    matchBuyOrder(incoming);
                } else {
                    matchSellOrder(incoming);
                }
            }
                
                
        public:
            void addOrder(Order order) {
                std::lock_guard<std::mutex> lock(mutex_);
                
                if (order.id <= 0 ||
                    order.price <= 0 ||
                    order.quantity <= 0
                ) {
                    return false;
                }

                if (orderIndex_.find(order.id) != orderIndex_.end()) {
                    return false;
                }

                matchOrder(order);

                if (order.quantity > 0) {
                    addToBook(order);
                }

                return true;
            }

            bool cancelOrder(std::int64_t orderId) {
                auto indexIt = orderIndex_.find(orderId);

                if (indexIt == orderIndex_.end()) {
                    return false;
                }

                OrderLocation location = indexIt->second;

                if (location.side == Side::Buy) {
                    auto priceIt = bids_.find(location.price);

                    priceIt->second.erase(location.iterator);

                    if (priceIt->second.empty()) {
                        bids_.erase(priceIt);
                    }
                } else {
                    auto priceIt = asks_.find(location.price);

                    priceIt->second.erase(location.iterator);

                    if (priceIt->second.empty()) {
                        asks_.erase(priceIt);
                    }
                }

                orderIndex_.erase(indexIt);

                return true;
            }

            bool modifyOrder(std::int64_t orderId, int newPrice, int newQuantity) {
                if (newPrice <= 0 || newQuantity <= 0) {
                    return false;
                }

                auto indexIt = orderIndex_.find(orderId);
                if (indexIt == orderIndex_.end()) {
                    return false;
                }

                Side side = indexIt->second.side;

                cancelOrder(orderId);

                Order newOrder{
                    orderId,
                    side,
                    newPrice,
                    newQuantity
                };

                return addOrder(newOrder);
            }


    };
